### PR TITLE
Derive conflict resolution

### PR DESCRIPTION
**Pull request by colin-grapl**
### Which issue does this PR correspond to?
#147 
#148

### What changes does this PR make to Grapl? Why?
1. Removes the underlying protobuf schemas for node types such as 'Process' and 'File', and instead leverages three distinct schemas - NodeDescription, IdentifiedNode, MergedNode.

2. Replaces 'string', 'int', and 'uint' with conflict resolvable types. The logic to enforce that resolution does not exist and will be implemented in a separate (and much smaller!) PR.

Specifically, while this diff appears very large, there are really just three places you need to worry about:

1. ` src/rust/derive-dynamic-node/src/lib.rs ` 
This is where the conflict resolution macro changes occur, allowing us to specify the resolution mechanism for a property.

2. ` src/proto/graplinc/grapl/api/graph/v1beta1/types.proto`
This is where the new proto structures are defined. The major changes are that we now have distinct node/graph types based on the state of the processed data (described, identified, merged) and, on top of that, the protobuf types to express the conflict resolution.

I've also documented most of the protobuf.

3. ` src/rust/graph-descriptions/src/lib.rs`
This is where the `impl` of these new protobuf types exist. The previously defined File, Process, etc are now in the `endpoint-plugin`.

All other changes are just to support this - there's a lot of diff noise involved, such as deleting tons of files that are no longer necessary, because the logic for nodes is now condensed.

The other churn is in the graph-merger, where I've spent some time refactoring the upsert logic, separating structures out into their own files. I've also added integration tests - 414 lines of additional integration tests are now in the graph-merger, as well as code to support those tests.

If you look at the git stats you'll find that while 104 files have been touched, the vast majority of those are just changing things like `Graph::new(0)` to `GraphDescription;:new()`, or that sort of thing. Also, apparently there have been some formatting changes, so that caused churn as well :(

### How were these changes tested?
I've tested locally and I've added lots of integration tests.

### CODE REVIEW

File: etc/local_grapl/bin/upload-sysmon-logs.py [hunk 1]
```diff
@@ -61,5 +61,5 @@ def parse_args():
             args.logfile,
             delay=args.delay,
             batch_size=args.batch_size,
-            use_links=args.use_links,
+            #use_links=args.use_links,
```
Comment (colin-grapl) [inline etc/local_grapl/bin/upload-sysmon-logs.py:64]: This was causing a crash. I need to follow up with Max about that.

File: src/rust/generators/osquery-subgraph-generator/src/tests/process.rs [hunk 2]
```diff
@@ -25,7 +25,7 @@ async fn test_subgraph_generation_process_create() {
         Err(Ok((subgraph, e))) => {
             assert!(
                 !subgraph.is_empty(),
-                "Generated subgraph was empty and errors were generated"
+                "Generated subgraphwas empty and errors were generated"
```
Comment (d0nut-grapl) [inline src/rust/generators/osquery-subgraph-generator/src/tests/process.rs:28]: lol wut

File: src/rust/generators/sysmon-subgraph-generator/src/models/process/create.rs [hunk 3]
```diff
@@ -82,7 +77,6 @@ pub fn generate_process_create_subgraph(
     );
 
     graph.add_edge("children", parent.clone_node_key(), child.clone_node_key());
-    graph.add_edge("parent", child.clone_node_key(), parent.clone_node_key());
```
Comment (d0nut-grapl) [inline src/rust/generators/sysmon-subgraph-generator/src/models/process/create.rs:85]: @colin-grapl is there a reason this was deleted?

**Review by d0nut-grapl — COMMENTED** [review]
Seems mostly the same from when we went over it (loving the new tests you added in btw).

I'd love for another pair of eyes to go over this btw just because it's so huge and I'm biased (saw once before :P )

File: src/rust/generators/sysmon-subgraph-generator/src/models/process/create.rs (hunk 3 above)
Comment (colin-grapl) [reply src/rust/generators/sysmon-subgraph-generator/src/models/process/create.rs:85]: It's redundant - there's no need to specify an edge and a reverse edge, one implies the other, and that's handled downstream.

File: src/rust/derive-dynamic-node/src/lib.rs [hunk 4]
```diff
@@ -15,16 +15,35 @@ use syn::{parse_quote,
...
-fn name_and_ty(field: &Field) -> (&Ident, &Type) {
-    (field.ident.as_ref().unwrap(), &field.ty)
+fn name_and_ty(field: &Field) -> (&Ident, &Type, String) {
+    let mut resolution = None;
+    for attr in &field.attrs {
+        on_grapl_attrs(&attr, |attr| {
+            match attr {
+                IMMUTABLE => resolution = Some(attr.to_string()),
+                INCREMENT => resolution = Some(attr.to_string()),
+                DECREMENT => resolution = Some(attr.to_string()),
```
Comment (christophermaier) [inline src/rust/derive-dynamic-node/src/lib.rs:32]: You could probably just use one match arm on `IMMUTABLE | INCREMENT | DECREMENT`.

File: src/rust/derive-dynamic-node/src/lib.rs [hunk 5]
```diff
@@ -55,105 +74,149 @@ pub fn derive_grapl_node(input: TokenStream) -> TokenStream {
     let node_trait_name = format!("I{}Node", struct_name);
     let node_trait_name = syn::Ident::new(&node_trait_name, struct_name.span());
 
+    let use_the_dead_code_method_name = format!("__avoid_dead_code_lint{}", struct_name);
```
Comment (christophermaier) [inline src/rust/derive-dynamic-node/src/lib.rs:77]: Should there be another `_` after "lint"?

That is, 
```rust
format!("__avoid_dead_code_lint_{}", struct_name);
```

File: src/rust/derive-dynamic-node/src/lib.rs [hunk 6]
```diff
@@ -166,64 +229,182 @@ pub fn derive_static_node_identity(input: TokenStream) -> TokenStream {
...
+    resolution_name: &str,
+) -> Option<(syn::Type, syn::Ident)> {
+    let (return_type, method_ident): (syn::Type, syn::Ident) = match property_type {
+        // janky way to get String="fully::qualified::path::Type" given a TypePath
+        Type::Path(typepath) => {
+            let typepath = typepath
+                .path
+                .segments
+                .iter()
+                .into_iter()
```
Comment (christophermaier) [inline src/rust/derive-dynamic-node/src/lib.rs:347]: `iter().into_iter()`??

File: src/proto/graplinc/grapl/api/graph/v1beta1/types.proto [hunk 7]
```diff
@@ -26,191 +34,146 @@ message IdStrategy {
...
-    string from = 1;
-    string to = 2;
-    // buf:lint:ignore FIELD_LOWER_SNAKE_CASE
-    string edgeName = 3;
+// Identified nodes are similar to a 'NodeDescription', except that their node_key is
+// canonical.
+message IdentifiedNode {
+    map<string, NodeProperty> properties = 1;
+    string node_key = 2;
+    string node_type = 3;
```
Comment (christophermaier) [inline src/proto/graplinc/grapl/api/graph/v1beta1/types.proto:112]: Can we go ahead and document _everything_ so we can enable that protobuf lint?

File: src/rust/Dockerfile [hunk 8]
```diff
@@ -6,6 +6,8 @@ SHELL ["/bin/bash", "-c"]
 
 RUN apt-get update && apt-get install -y --no-install-recommends \
         wait-for-it \
+        ca-certificates \
+        netbase \
```
Comment (inickles-grapl) [inline src/rust/Dockerfile:10]: These seem unnecessary for the build image, we already have them in the deploy base. Can these be removed?

**Review by christophermaier — COMMENTED** [review]
Just a few minor comments... still not as familiar with this code as I feel I would need to be to be more constructive.

File: .gitignore [hunk 9]
```diff
@@ -1,6 +1,7 @@
 # Mac OS
 .DS_Store
-
+.ninja_log
+.pytype/
```
Comment (inickles-grapl) [inline .gitignore:4]: Do these belong under "Mac OS"? Seems doubtful.

File: src/rust/endpoint-plugin/src/graph.rs [hunk 10]
```diff
@@ -0,0 +1 @@
+
```
Comment (inickles-grapl) [inline src/rust/endpoint-plugin/src/graph.rs:1]: Empty file?

File: src/rust/endpoint-plugin/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+                                              ProcessOutboundConnection,
+                                              ProcessOutboundConnectionNode}};
+
+#[cfg(test)]
+mod tests {
+    #[test]
+    fn it_works() {
+        assert_eq!(2 + 2, 4);
+    }
+}
```
Comment (inickles-grapl) [inline src/rust/endpoint-plugin/src/lib.rs:48]: This test is fake. Should either be removed or replaced with a real one.

File: src/rust/generators/sysmon-subgraph-generator/src/models/process/create.rs [hunk 12]
```diff
@@ -91,3 +85,66 @@ pub fn generate_process_create_subgraph(
...
...Data Name='ProcessGuid'>{87E8D3BD-9DDA-5D38-0000-0010A3941D00}</Data><Data Name='ProcessId'>5752</Data><Data Name='Image'>C:\Windows\System32\cmd.exe</Data><Data Name='FileVersion'>10.0.10240.16384 (th1.150709-1700)</Data><Data Name='Description'>Windows Command Processor</Data><Data Name='Product'>Microsoft� Windows� Operating System</Data><Data Name='Company'>Microsoft Corporation</Data><Data Name='OriginalFileName'>Cmd.Exe</Data><Data Name='CommandLine'>"cmd" /C "msiexec /quiet /i cmd.msi"</Data><Data Name='CurrentDirectory'>C:\Users\grapltest\Downloads\</Data><Data Name='User'>DESKTOP-FVSHABR\grapltest</Data><Data Name='LogonGuid'>{87E8D3BD-99C8-5D38-0000-002088140200}</Data><Data Name='LogonId'>0x21488</Data><Data Name='TerminalSessionId'>1</Data><Data Name='IntegrityLevel'>Medium</Data><Data Name='Hashes'>MD5=A6177D080759CF4A03EF837A38F62401,SHA256=79D1FFABDD7841D9043D4DDF1F93721BCD35D823614411FD4EAB5D2C16A86F35</Data><Data Name='ParentProcessGuid'>{87E8D3BD-9DD8-5D38-0000-00109F871D00}</Data><Data Name='ParentProcessId'>6132</Data><Data Name='ParentImage'>C:\Users\grapltest\Downloads\svchost.exe</Data><Data Name='ParentCommandLine'>.\svchost.exe</Data></EventData></Event>"#;
```
Comment (inickles-grapl) [inline src/rust/generators/sysmon-subgraph-generator/src/models/process/create.rs:117]: I think this should just be in a file under a "testdata" dir or something.

File: src/rust/grapl-service/src/serialization/zstd_proto_graph.rs [hunk 13]
```diff
@@ -58,7 +58,93 @@ impl CompletionEventSerializer for SubgraphSerializer {
             return Ok(vec![]);
         }
 
-        // TODO: This should be a monitor
+        for edge_list in subgraph.edges.values_mut() {
+            edge_list.edges.sort_unstable();
+            edge_list.edges.dedup();
+        }
+
+        // TODO: This should be a metric
```
Comment (inickles-grapl) [inline src/rust/grapl-service/src/serialization/zstd_proto_graph.rs:66]: Should this TODO be implemented in this PR?

**Review by inickles-grapl — CHANGES_REQUESTED** [review]
I have questions on a couple items.

File: src/rust/grapl-service/src/serialization/zstd_proto_graph.rs (hunk 13 above)
Comment (colin-grapl) [reply src/rust/grapl-service/src/serialization/zstd_proto_graph.rs:66]: I don't think so, it wasn't introduced in this PR

File: src/rust/generators/sysmon-subgraph-generator/src/models/process/create.rs (hunk 12 above)
Comment (colin-grapl) [reply src/rust/generators/sysmon-subgraph-generator/src/models/process/create.rs:117]: I'd rather handle that in another PR if that's alright?

**Review by inickles-grapl — APPROVED** [review]

**Pull request merged**